drivers/hmis/app/graphql/mutations/update_external_form_submission.rb

```diff
@@ -8,24 +8,72 @@ module Mutations
   class UpdateExternalFormSubmission < CleanBaseMutation
     argument :input, Types::HmisSchema::ExternalFormSubmissionInput, required: true
     argument :id, ID, required: true
+    argument :project_id, ID, required: false # not required for backwards compatibility
 
-    field :external_form_submission, Types::HmisSchema::ExternalFormSubmission, null: false
+    field :external_form_submission, Types::HmisSchema::ExternalFormSubmission, null: true
     field :errors, [Types::HmisSchema::ValidationError], null: false, resolver: Resolvers::ValidationErrors
 
-    def resolve(id:, input:)
+    def resolve(...)
+      Hmis::Hud::Base.transaction do
+        _resolve(...)
+      end
+    end
+
+    protected
+
+    EXTRANEOUS_KEYS = ['captcha_score', 'form_definition_id', 'form_content_digest'].freeze
+
+    def _resolve(id:, project_id:, input:)
       record = HmisExternalApis::ExternalForms::FormSubmission.find(id)
       raise 'Access denied' unless allowed?(permissions: [:can_manage_external_form_submissions])
 
       record.assign_attributes(**input.to_params)
 
-      errors = []
-      if record.valid?
-        record.save!
-      else
-        errors = record.errors
-        record = nil
+      errors = HmisErrors::Errors.new
+
+      unless record.valid?
+        errors.add_ar_errors(record.errors&.errors)
+        return { errors: errors }
       end
 
+      if record.status_changed? && record.status == 'reviewed' && !record.spam
```

Comment:
We will need to refactor some of this out into shared code when we add a mutation for "bulk review". Thinking ahead to that, would it be possible to refactor out most of this processing code to live on the FormSubmission record?

```ruby
if record.status_changed? && record.status == 'reviewed' && !record.spam
   record.run_form_processor(project: project) #<== processes and build related records, but doesnt save

   if record.enrollment&.new_record?
          record.enrollment.client.save!
          record.enrollment.save_new_enrollment!
    end
   record.form_processor.save!
end
```

Reply:
Yep I think it looks good. I'm on the fence on whether it should be `creates_client_or_enrollment` or `updates_client_or_enrollment`... the presence of the "mapping" doesn't tell us whether the form is used for creation or updating (And sometimes the same form is used for both). The link_id_item_hash thing is on an instance variable so we should only need to traverse the definition once